Cleanup Complex

hughbe opened the pull request:
- Format the source file
- Combine tests into one source file
- Rewrite tests to make advantage of xunit

mellinoe commented:
The first commit looks good to me. Definitely much more readable and organized. It will take me a bit longer to get through the changes in the tests.

In the meantime, could you fix the typos in the commit messages?

Cleanup **context** tests
Cleanup and format Complex source **fiel**

hughbe commented:
I've modified the names and followed up on your PR feedback.

Somehow, I've broken all the Unix builds. I'll take a look soon but I'm a bit busy atm - if you see the cause of that do drop a comment identifying the issue! :)

mellinoe commented:
It seems that the tests are actually failing on non-Windows platforms. Do you have a VM you could test them in? You can see the failure messages here:

http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/ubuntu_debug_tst_prtest/1638/console

It looks like some `Exp`, `Equals`, and `Log` tests are failing. We should try to improve the assertion messages along with this change so that a log such as the above one immediately points out the issue.

mellinoe commented:
I'm reviewing through the test changes now, looking at it in Visual Studio since it's easier to analyze the diff there.

hughbe commented:
I don't have a VM for this atm, I'll try to get one tomorrow but if I have time! Meanwhile if you find something that I've accidently changed, please point it out 
thanks

mellinoe commented:
@hughbe, the Complex test file is too big for me to make comments on directly, so I'll just comment in one big blob here, because I have my notes in a document anyways since I reviewed from VS.

[Line 111](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L111) I'd move the comment to above each "block" and keep the mention of the quadrant: // First quadrant: +, + // Second quadrant: +, - // Third, etc.

[Line 611](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L611) I think this is supposed to be "Random values", not "Invalid values". The invalid cases are just below.

[Line 612](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L612) Could you do similar to above and keep the mention of quadrants (and move the comments above each `yield` statement. Also, could you double-check that we actually cover all of the quadrant cases here? I see that in the original test code we had duplicated the -, +, -, + test case, but the comment for the duplicated one seems to indicate that it should have been a test case for the 4th-quadrant / 4th-quadrant combination, which would be +, -, +, -. I don't think we have a test case for that currently.

[Line 1122](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L1122) I wonder if we really need to keep this loop? It seems sort of random to force extra test cases for negate, but not for any of the other methods. A bit of a nit-pick, though; I doubt this has any effect on run-time or anything of that sort.

[Line 74](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L74) I guess here's another instance of the loop. I'm not sure it's even valuable here since the constructor is so trivial.

[Line 1676](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L1676) Just wanted to point out that the original code makes some mention of a `germanNumberFormatInfo`, but it actually uses the current culture's `NumberFormat` (like yours does). I wonder if we originally were trying to test the German number format info but lost that in a port somewhere a long time ago? Is there something particular in the German number format that would be valuable to these tests?

[Line 770](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L770) I think this method is mis-named, or something of the sort. It doesn't have test data, it has assertions in it.

[Line 775](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L775) This will be very hard to diagnose for failures, because everything will just say "Excepted: true, Actual: false" or something like that. We should add string messages to all of the places that we use Assert.Equal(bool, bool). The original test cases had some messages so they were a bit more easily diagnosable. For example: `string.Format("c1:{0} == c2{1} is not '{2}' as expected", c1, c2, expectedResult)`. The messages don't need to be as verbose as the original, but I think we need something.

[Line 816](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L816) Where did this code at the top come from (if and else-if blocks)? I don't see it in the original test cases, which seem to be otherwise identical.

[Line 955](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L955) Unused method?

[Line 44](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L44) `Assert.Equal(0, Complex.Zero)`

[Line 47](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L47) There are other test cases with things like `number * Complex.Zero == 0`, `number - Complex.Zero == number`, etc. I think we should keep those. They can probably just go in the `Zero()` method itself. Unless I have missed them somewhere else. Also applies to `One`, `ImaginaryOne`, etc.

[Line 320](https://github.com/hughbe/corefx/blob/45d0e74e57d76e638e2ffc7b769fd255d88159d1/src/System.Runtime.Numerics/tests/ComplexTests.cs#L320) This doesn’t seem to be the same as the old edge case:
````CSharp
// Verify edge cases where imaginary part is zero and real part is negative.
VerifyASin(-1234000000, 0, -1.5707963267948966, 21.62667394298955);
````
General comments:
* I wonder if it wouldn't be more readable to use something like `SmallPositiveDouble()` `SmallNegativeDouble()` instead of `SmallRandomDouble(false)` `SmallrandomDouble(true)`. I know this is how the original tests were written, but I find it difficult to automatically associate `false` with "positive number" when I'm skimming through the tests. You should be able to try this out with a quick Find-And-Replace and see if it looks better to you. I have a feeling it would make some of the test cases more obvious, like all of the negative/positive combination blocks of code.

* I think adding back the "quadrant" comments would be nice, we can probably remove the "-, -, +, +" comments if we do the above helper method rename.

* VerifyRealImaginaryProperties should have a better failure message. It's too hard to diagnose the error if you don't see the expected/actual value differences. Right now we only get the input data, which we should already have since we know which xUnit test case failed. Same applies to VerifyMagnitudePhaseProperties.

* I wonder if we can pull out some common test data. For example, it seems like a lot of the TestData methods have the same quadrant combinations. Unary operators/methods seem to have many of the same combinations (1/0 combos, +/- combos), etc. We could probably pull those things out into helper enumerators and then yield them from TestData methods as needed. This might be going a bit too far in obscuring where the test data is coming from, though. If you do end up doing this, please add it into a separate commit so I can diff it easily.

* I don't see any Square root tests. Am I missing them somewhere?

hughbe commented:
@mellinoe Thanks for that review - I've done most things in the latest PR - hopefully this will help me fix the *nix tests, if my current attempts have failed.

I've addressed and corrected all of the following:
- Line 1676, r.e. German number format: the only thing that I can think of is the use of `.` instead of `,` for thousands, and `,` instead of `.` for decimals. Arguably, there is no point testing this behavior in this project, as System.Globalization.* tests cover this.

- Line 47 r.e. tests for Complex.Zero, Complex.One etc.: I've subsumed these into the `Multiply`, `Subtract`, `Divide` and `Add` tests (see right at the top of the `*_TestData` iterators for these tests)

- Line 816 r.e. if and else-if: this is actually what was breaking part of the *nix build, but weirdly passes on Windows, I think. This code was added to special-case the complex `{double.MaxValue, 0}` and `{double.MaxValue, double.MaxValue}`

- General comment on condensing common test data: definitely a good idea. I like this, and shall do this after you've reviewed the latest commits and the build failures are fixed :)

hughbe commented:
Grr, GitHub won't show the diff for the latest commit - it doesn't seem too long! Sorry if you have to go through the fuss of downloading and reviewing in VS!

mellinoe commented:
I looked through the latest version and the changes look good to me. Thanks for adding back the Sqrt tests.

> Line 1676, r.e. German number format: the only thing that I can think of is the use of . instead of , for thousands, and , instead of . for decimals. Arguably, there is no point testing this behavior in this project, as System.Globalization.* tests cover this.

Sounds reasonable. Perhaps it's related to the fact that the ToString method already puts a comma in between the real/imaginary parts, and we wanted to check that the formatting still worked for some reason. My care factor is low here because the tests were already not using a German NumberInfo (the variable name just indicated that it was).

> I've subsumed these into the Multiply, Subtract, Divide and Add tests

Yep, see it now, thanks

> Line 816 r.e. if and else-if: this is actually what was breaking part of the *nix build, but weirdly passes on Windows, I think. This code was added to special-case the complex {double.MaxValue, 0} and {double.MaxValue, double.MaxValue}

Interesting, it seems like this will be the key to figuring out the test failures. This was the only thing in the PR that jumped out at me as being functionally different from the old version of the tests. Perhaps you can trace through the old tests and see why they don't need this workaround on Windows.

> General comment on condensing common test data: definitely a good idea. I like this, and shall do this after you've reviewed the latest commits and the build failures are fixed :)

Feel free to leave this for a separate PR, I would be fine with the current state of things assuming we can figure out that workaround issue above. This would be additional goodness on top of the current cleanup.

hughbe commented:
@mellinoe 
Here is the test that I was translating: https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Numerics/tests/Complex/standardNumericalFunctions_Exp.cs#L93

Seems like there was a typo here. It isn't actually testing the complex {double.MaxValue, 0}, but actually the complex {double.MaxValue, double.MaxValue} again!

What I think is the case is that performing Complex.Exp(new Complex(double.MaxValue, 0)) gives undefined results - i.e. it depends on the platform?

mellinoe commented:
> Seems like there was a typo here. It isn't actually testing the complex {double.MaxValue, 0}, but actually the complex {double.MaxValue, double.MaxValue} again!

Interesting, it looks like you're right. The `maxReal` variable is unused and the next line just uses `max` again (even though that exact same thing was done a few lines earlier). Since this was previously untested, we can probably add this new test case to a separate method and mark it with an ActiveIssue if we can't figure out why it's failing. It may not be a valid test case.

`Complex.Exp` just uses `Math.Exp`, `Math.Sin`, and `Math.Cos`. Perhaps there is different behavior there in this edge case.

hughbe commented:
How interesting. Made issue #6022 for discussion of this, and pushed a commit

hughbe commented:
More inconsistencies - Ubuntu builds are failing. *Thankfully* only on one test:
This seems to be at Log (Complex in 1st quadrant, Complex in 2nd quadrant). Here's the log:
 http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/opensuse13.2_debug_tst_prtest/1104/consoleFull#-21371504001f1a4601-6aec-4fd5-b678-78d4389fd5e8

mellinoe commented:
For this kind of thing, I would recommend just trying to narrow down why the test wasn't failing before, and trace through the steps of the test, perhaps printing some stuff out to see intermediate values. You may be able to pinpoint where the values diverge.

hughbe commented:
http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/opensuse13.2_debug_tst_prtest/1104/consoleFull#-21371504001f1a4601-6aec-4fd5-b678-78d4389fd5e8 is what's really really confusing me.

Take a look at 14:53:50 onwards.
Here's the code that we're testing: https://github.com/hughbe/corefx/blob/numerics-cleanup/src/System.Runtime.Numerics/tests/ComplexTests.cs#L819

I just don't understand why this is failing - there seems to be some sort of strange manipulation of the data for the equals tests. Have we encountered some sort of strange CLR bug?!

mellinoe commented:
````
c1:(0, -6.20034545379439E-152) == c2(0, 0) is not 'True' as expected
````

One thing that springs to mind is that your refactoring might have changed the way some of the values are passed around through these validation methods/helpers. I recall some issues in the past where test values were moved around to different intermediate memory locations (rather than stay in a high-precision register) due to small, seemingly unrelated changes, which showed up as test failures sort of like this. For example, I remember that some of the refactoring you did changed a Complex parameter to a real-imaginary pair of parameters to avoid having two separate overloads. Also, the test values are produced from a separate method now (the _TestData methods) and stored in arrays, which xUnit passes into the method through some mechanism.

I haven't actually validated any of these ideas, but perhaps some of the above is changing how these parameters are getting stored / passed around on non-Windows systems and that is affecting their intermediate precision, compounding into these small differences in value. Like I said, I think tracing through and following the intermediate values would be a lot of help in diagnosing things.

hughbe commented:
@mellinoe 
I don't have an OpenSuse/CentOS/Ubuntu/Mac OSX VM on hand (sorry :P), but I pushed a fix that hopefully will fix the failures if your theory is correct.
I'm also hoping that the CI starts working soon so that I can see how its working

mellinoe commented:
@hughbe With the latest patch, all of the tests except for one pass on my Ubuntu machine. I haven't looked too far through your code yet to see what you changed but it seems like you just removed the use of `MemberData`, which might confirm what I suspected above. I'll look into the last few patches once we fix the last test issue.

I'm still seeing one failure:

````
     System.Numerics.Tests.cast_fromTest.RunSingleExplicitCastFromBigIntegerTests [FAIL]
        System.OverflowException : BigInteger cannot represent infinity.
        Stack Trace:
              at System.Numerics.BigInteger..ctor(Double value)
              at System.Numerics.BigInteger.op_Explicit(Double value)
              at System.Numerics.Tests.cast_fromTest.GenerateRandomBigIntegerGreaterThan(Double value, Random random)
              at System.Numerics.Tests.cast_fromT
````

hughbe commented:
That doesn't seem to be caused by anything I changed? Ahh I wasn't expecting this PR to turn out such weird stuff. I'll check it out

stephentoub commented:
@dotnet-bot test this please

stephentoub commented:
@dotnet-bot test this please

hughbe commented:
OpenSuse tests are passing now (well at least for System.Runtime.Numerics.Tests ;) :
 System.Runtime.Numerics.Tests  Total: 1118, Errors: 0, Failed: 0, Skipped: 0, Time: 12.515s

stephentoub commented on src/System.Runtime.Numerics/src/System/Numerics/Complex.cs:
```diff
@@ -2,84 +2,41 @@
+    /// <summary>
+    /// A complex number z is a number of the form z = x + yi, where x and y 
+    /// are real numbers, and i is the imaginary unit, with the property i2= -1.
+    /// </summary>
```
Where did this description come from?  I get that there are only so many ways to describe what is a complex number, but this is almost word for word the summary on wikipedia (https://en.wikipedia.org/wiki/Complex_number).

stephentoub replied:
FYI, the plan is that very soon the MSDN docs will be used to populate XML comments in the source.  It's not worth spending time adding such comments manually; in fact, doing so could cause the other process to take longer.

hughbe replied:
Comment came from the massive file header at the top (that I removed to combine with the class summary)

Sorry about the doc changes - It's probably not worth the bother to undo that but I'll keep it in mind for the future

stephentoub replied:
Ok.

stephentoub commented on src/System.Runtime.Numerics/src/System/Numerics/Complex.cs:
```diff
@@ -185,138 +135,73 @@ public static Double Abs(Complex value)
-        public bool Equals(Complex value)
-        {
-            return ((_real.Equals(value._real)) && (_imaginary.Equals(value._imaginary)));
+            return Equals((Complex)obj);
```
Why did you change this?

hughbe replied:
Its a form of refactoring - once we know that the user passed a Complex as the object parameter, we can eliminate identical code, preventing bugs and enhancing readability :)

JonHanna replied:
It's not identical code, so it's not just a refactoring. The observable change here should probably be considered a bug fix, in that `Equals(object)` should probably behave the same as `Equals(Complex)` if that differs from `==`, which this does if either the real or imaginary part is NaN, especially since the current implementation is broken for non-generic hash-tables or dictionaries that cast it to `object`. It should probably be separate in case there's some reason why the behaviour can't be changed.

hughbe replied:
Oh right I'll undo this

JonHanna replied:
We should definitely have an issue to track it, and let you say you'll fix it after this PR (so nobody else fixes it and forces you to rebase again).

JonHanna replied:
@hughbe Did you do anything about this separately? This may not have been the place to fix a bug, but you'd still found and fixed a bug all the same; `((object)a).Equals((object)a)` should always be true.

hughbe replied:
Oh no, I didn't, sorry. I'll file an issue

hughbe replied:
@JonHanna 
Filed issue #6338 to fix this.
I will send a PR request if the potential breaking change effect of the bug is outweighed by the benefits of consistency.

stephentoub commented on src/System.Runtime.Numerics/src/System/Numerics/Complex.cs:
```diff
@@ -185,138 +135,73 @@ public static Double Abs(Complex value)
-
-        // --------------SECTION: Trigonometric operations (methods implementing ITrigonometric)  -------------- //
-
+        
```
Please revert these changes to ToString.  From my perspective, these changes make it harder to understand the behavior of most of the overloads, and it can actually add some cost.

JonHanna commented:
Do the tests have to be in one single file? 2,271 lines is quite large.

They're also all in one class. Having a separate class for e.g. arithmetic and a separate for trig, and so on would mean more concurrency if one is just running these tests, since XUnit runs each class concurrently, but tests within a class sequentially. It won't make much difference if someone is running all the corefx tests since there'll be so many other tests to run, but could if they're just running these and have cores to spare.

hughbe commented:
Jon, there seems to be a trend in the repo to combine multiple test files into a single one - see the recent System.Collections/System.Collectiobs.Specialized tests work. I'm just following what I see as a more clean and sensible way of collating tests.

JonHanna commented:
You've definitely got it much cleaner than it is before. I just wonder if the sweet-spot between too many files and too-large files is a slightly larger chunk.

hughbe commented:
Thats a good point. I think, however, there's quite a lot of common test data in the TestData iterators for each function. As Eric suggested, a future PR could consolidate this data, removing a considerable amount of code.

As ever, your wish is my command. If you think its a good idea to split up the files, then i shall abide. If you do decide on this, then maybe it could be a good idea to split up the files by similar functions, not just single functions (e.g. trigonometry, algebra, casting, constructors)

hughbe commented:
I just pushed a PR feedback commit. I merged it with a smaller Fix non-Windows test failures commit so that I don't end up spamming the commit history!

JonHanna commented:
Nope, your opinion has more weight than mine here: I haven't found an actual flaw or bug, I'm just another community member submitting issues and PRs like yourself, and the tie-breaker is that you're the one putting in the work :+1:

mellinoe commented:
The latest changes look good to me, and the tests are passing on my Linux machine and in my VM. @stephentoub Any other concerns? The Equals and ToString changes have been reverted.

I'd still like to get some additional refactoring and consolidation done (like I mentioned a bit above), but we can try to tackle that separately. I think the current state is much cleaner than before.

stephentoub commented:
> @stephentoub Any other concerns?

The product changes look ok.

mellinoe commented:
Thanks for the contribution, @hughbe . The code is much more in line with our guidelines now. There's a few more things to be done, mainly around organizing and reusing common pieces of the code better, but this is a good improvement for now.
